python/instrumentation/openinference-instrumentation-langchain/tests/test_instrumentor.py

```diff
@@ -92,6 +95,72 @@ async def f(_: Any) -> Optional[Span]:
     }
 
 
+async def test_get_current_chain_root_span(
+    in_memory_span_exporter: InMemorySpanExporter,
+) -> None:
+    """Test retrieving the current chain root span during RunnableLambda execution."""
+    n = 10  # Number of concurrent runs
+    loop = asyncio.get_running_loop()
+
+    root_spans_during_execution = []
+
+    def f(x: int) -> int:
+        root_span = get_current_root_chain_span()
+        assert root_span is not None, "Root span should not be None during execution (sync)"
+        root_spans_during_execution.append(root_span)
+        return x + 1
+
+    with ThreadPoolExecutor() as executor:
+        tasks = [loop.run_in_executor(executor, RunnableLambda(f).invoke, 1) for _ in range(n)]
+        await asyncio.gather(*tasks)
+
+    root_span_after_execution = get_current_root_chain_span()
+    assert root_span_after_execution is None, "Root span should be None after execution"
+
+    assert len(root_spans_during_execution) == n, "Did not capture all root spans during execution"
+
+    spans = in_memory_span_exporter.get_finished_spans()
+    assert len(spans) == n, f"Expected {n} spans, but found {len(spans)}"
+
+
+async def test_get_current_chain_root_span_async(
+    in_memory_span_exporter: InMemorySpanExporter,
+) -> None:
+    """Test retrieving the current chain root span during RunnableLambda execution."""
+    if sys.version_info < (3, 11):
+        pytest.xfail("Async test may fail on Python versions below 3.11")
+    n = 10  # Number of concurrent runs
+
+    root_spans_during_execution = []
+
+    async def f(x: int) -> int:
+        root_span = get_current_root_chain_span()
+        assert root_span is not None, "Root span should not be None during execution (async)"
+        root_spans_during_execution.append(root_span)
+        return x + 1
+
+    sequence: RunnableSerializable[int, int] = RunnableLambda[int, int](f) | RunnableLambda[
+        int, int
+    ](f)
+
+    for _ in range(n):
+        await sequence.ainvoke(1)
+
+    root_span_after_execution = get_current_root_chain_span()
+    assert root_span_after_execution is None, "Root span should be None after execution"
+
+    assert (
+        len(root_spans_during_execution) == 2 * n
+    ), "Did not capture all root spans during execution"
+
+    assert (
+        len(set(id(span) for span in root_spans_during_execution)) == 2 * n
```

Comment:
The RunnableLambdas create their own chain, so they are their own root

Reply:
> The RunnableLambdas create their own chain, so they are their own root

If that's the case then this is not really testing the tree climbing procedure. Is that correct?

Reply:
It's not, though if like you suggested we yield the the ancestor chains we can test how many there are. I didn't end up doing that since it feels like a confusing interface?

Reply:
This is not related to yielding the ancestor chain. All i'm pointing out is that we are _not_ testing what we have implemented.

Reply:
I know that, I'm suggesting yielding the ancestor chain can be an easy way to do that

Reply:
> I know that

So are you suggesting that the logic you have implemented so far doesn't need to be tested?

Reply:
I’m not saying it has to be tested, but I’ve noticed the logic isn’t being tested. Since we seem to agree on that observation, I’d like to understand why leaving it untested is considered the better option.

Reply:
I am specifically offering an alternative strategy to testing whether or not it's the right ancestor directly: if we change this implementation to yield ancestors we can do (I think) a good enough test by counting how many ancestors are yielded each time instead of directly checking the ancestor directly.

Regarding needing to test it: I can go either way on it, I'm not positive this test is needed though I agree it would be nice. I'm using tests as a guide here to scaffold the implementation, not necessarily be exhaustive.

Reply:
> I'm not positive this test is needed though I agree it would be nice.  I'm using tests as a guide here to scaffold the implementation, not necessarily be exhaustive.

While I understand the test may not seem essential right now, it’s important to consider that this repo is updated only occasionally. Given the infrequent changes and the high likelihood that any future updates may be handled by someone new, having tests that are both thorough and precise can be incredibly beneficial. These tests can serve as reliable documentation and help prevent potential issues when the codebase is revisited later.

> I am specifically offering an alternative strategy to testing 

As an aside, I previously mentioned that the self.run_map from the base instrumentor already stores the family tree of UUIDs. You can leverage this to simplify both the implementation and the testing, as it’s the ultimate source of truth for the family tree.